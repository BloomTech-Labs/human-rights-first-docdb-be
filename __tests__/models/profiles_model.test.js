const db = require('../../data/db-config');
const Profiles = require('../../api/profile/profileModel');

beforeAll(async () => {
  await db.migrate.rollback();
  await db.migrate.latest();
});

beforeEach(async () => {
  await db.seed.run();
});

afterAll(async () => {
  await db.destroy();
});

describe('findAll()', () => {
  let res;
  beforeEach(async () => {
    res = await Profiles.findAll();
  });

  it('returns 8 profiles', () => {
    expect(res).toHaveLength(8);
  });

  it('returns the correct data shape', () => {
    expect(res).toMatchSnapshot();
  });
});

describe('findById(id)', () => {
  let res;
  const id = 'abc123';
  beforeEach(async () => {
    res = await Profiles.findById(id);
  });
  it('returns correct profile by id', () => {
    expect(res.name).toBe('John Smith');
    expect(res.email).toBe('abc@gmail.com');
  });
  it('returns the correct data shape', () => {
    expect(res).toMatchSnapshot();
  });
});

describe('create(profile)', () => {
  let res;
  const newProfile = {
    id: '0630SM',
    email: 'rabbit@moon.com',
    name: 'Usagi Tsukino',
    avatarUrl: 'https://tinyurl.com/y7adbfe8',
  };

  beforeEach(async () => {
    res = await Profiles.create(newProfile);
  });

  it('returns newly created profile', () => {
    expect(res[0]).toMatchObject(newProfile);
  });

  it('returns 9 profiles', async () => {
    const getAll = await db('profiles');

    expect(getAll).toHaveLength(9);
    expect(getAll[8]).toMatchObject(newProfile);
  });
  it('returns the correct data shape', () => {
    expect(res).toMatchSnapshot();
  });
});

describe('update(id, profile)', () => {
  let res;
  const oldProfile = {
    id: 'bnsdj2',
    email: 'Rei@gmail.com',
    name: 'Rei Hino',
    avatarUrl: 'https://tinyurl.com/4pf7d3v6',
  };

  const updatedProfile = {
    id: 'bnsdj2',
    email: 'Minako@gmail.com',
    name: 'Minako Aino',
    avatarUrl: 'https://tinyurl.com/4pf7d3v6',
  };
  beforeEach(async () => {
    res = await Profiles.update(oldProfile.id, updatedProfile);
  });
  it('finds and returns correct profile with updated information', async () => {
    const findById = await db('profiles')
      .where({ id: 'bnsdj2' })
      .select('*')
      .first();
    expect(res[0].name).toBe('Minako Aino');
    expect(findById).toMatchObject(updatedProfile);
  });
  it('returns the correct data shape', () => {
    expect(res).toMatchSnapshot();
  });
});

describe('remove(id)', () => {
  const profileToRemove = {
    id: 'bnsdj2',
    email: 'Rei@gmail.com',
    name: 'Rei Hino',
    avatarUrl: 'https://tinyurl.com/4pf7d3v6',
  };
  it('removes a profile from the db', async () => {
    await Profiles.remove(profileToRemove.id);
    const findAll = await db('profiles');
    expect(findAll).toHaveLength(7);
  });
  it('finds and removes the correct profile', async () => {
    const findProfileBeforeRemove = await db('profiles')
      .where({ id: profileToRemove.id })
      .select('*')
      .first();
    expect(findProfileBeforeRemove.name).toBe('Rei Hino');

    await Profiles.remove(profileToRemove.id);

    const findProfileAfterRemoval = await db('profiles')
      .where({ id: profileToRemove.id })
      .select('*')
      .first();

    expect(findProfileAfterRemoval).not.toBeDefined();
  });

  it('returns the correct data shape', async () => {
    let res = await Profiles.remove(profileToRemove.id);

    expect(res).toBe(1);
  });
});

describe('findOrCreateProfile()', () => {
  let res;
  const existingProfile = {
    id: 'bnsdj2',
    email: 'Rei@gmail.com',
    name: 'Rei Hino',
    avatarUrl: 'https://tinyurl.com/4pf7d3v6',
  };

  const newProfile = {
    id: '0630SM',
    email: 'rabbit@moon.com',
    name: 'Usagi Tsukino',
    avatarUrl: 'https://tinyurl.com/y7adbfe8',
  };

  it('finds the profile in the database if it exists', async () => {
    const findProfile = await db('profiles')
      .where({ id: 'bnsdj2' })
      .first()
      .select('*');
    expect(findProfile.name).toBe('Rei Hino');

    res = await Profiles.findOrCreateProfile(existingProfile);
    expect(res).toMatchObject(findProfile);
  });
  it('creates a new profile in the database if the profile did not exist', async () => {
    let noProfile = await db('profiles')
      .where({ id: '0630SM' })
      .first()
      .select('*');
    expect(noProfile).not.toBeDefined();

    res = await Profiles.findOrCreateProfile(newProfile);
    expect(res).toMatchObject(newProfile);

    const allProfiles = await db('profiles');
    expect(allProfiles).toHaveLength(9);
  });
  it('returns the correct data shape', async () => {
    res = await Profiles.findOrCreateProfile(newProfile);
    expect(res).toMatchSnapshot();
  });
});
