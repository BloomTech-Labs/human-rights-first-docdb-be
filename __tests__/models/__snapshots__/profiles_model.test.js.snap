// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create(profile) returns the correct data shape 1`] = `
Array [
  Object {
    "avatarUrl": "https://tinyurl.com/y7adbfe8",
    "email": "rabbit@moon.com",
    "id": "0630SM",
    "name": "Usagi Tsukino",
  },
]
`;

exports[`findAll() returns the correct data shape 1`] = `
Array [
  Object {
    "avatarUrl": "tinyurl.com/2p8bsvbz",
    "email": "abc@gmail.com",
    "id": "abc123",
    "name": "John Smith",
  },
  Object {
    "avatarUrl": "tinyurl.com/mph27afd",
    "email": "mary@gmail.com",
    "id": "2h38dh",
    "name": "Mary Berry",
  },
  Object {
    "avatarUrl": "https://tinyurl.com/4ffmpmxu",
    "email": "brian@gmail.com",
    "id": "j3iasd",
    "name": "Brian Cranston",
  },
  Object {
    "avatarUrl": "https://tinyurl.com/2p9a9bmm",
    "email": "jessica@hotmail.com",
    "id": "23jsbb",
    "name": "Jessica Rabbit",
  },
  Object {
    "avatarUrl": "https://tinyurl.com/ycktu39j",
    "email": "Santa@gmail.com",
    "id": "jsh123",
    "name": "Santa Claus",
  },
  Object {
    "avatarUrl": "https://tinyurl.com/4pf7d3v6",
    "email": "Rei@gmail.com",
    "id": "bnsdj2",
    "name": "Rei Hino",
  },
  Object {
    "avatarUrl": "https://tinyurl.com/2hy5at25",
    "email": "jpeck@gmail.com",
    "id": "jpe12f",
    "name": "Josh Peck",
  },
  Object {
    "avatarUrl": "https://tinyurl.com/488ef4xm",
    "email": "test_user@gmail.com",
    "id": "00ultx74kMUmEW8054x6",
    "name": "003",
  },
]
`;

exports[`findById(id) returns the correct data shape 1`] = `
Object {
  "avatarUrl": "tinyurl.com/2p8bsvbz",
  "email": "abc@gmail.com",
  "id": "abc123",
  "name": "John Smith",
}
`;

exports[`findOrCreateProfile() returns the correct data shape 1`] = `
Object {
  "avatarUrl": "https://tinyurl.com/y7adbfe8",
  "email": "rabbit@moon.com",
  "id": "0630SM",
  "name": "Usagi Tsukino",
}
`;

exports[`update(id, profile) returns the correct data shape 1`] = `
Array [
  Object {
    "avatarUrl": "https://tinyurl.com/4pf7d3v6",
    "email": "Minako@gmail.com",
    "id": "bnsdj2",
    "name": "Minako Aino",
  },
]
`;
